### PR TITLE
DSA mod inverse fix.

### PR DESCRIPTION
**Pull request by paulidale**
There is a side channel attack against the division used to calculate one of
the modulo inverses in the DSA algorithm.  This change takes advantage of the
primality of the modulo and Fermat's little theorem to calculate the inverse
without leaking information.

Thanks to Samuel Weiser for finding and reporting this.

### CODE REVIEW

File: crypto/dsa/dsa_ossl.c [hunk 1]
```diff
@@ -259,7 +260,7 @@ static int dsa_sign_setup(DSA *dsa, BN_CTX *ctx_in,
         goto err;
 
     /* Compute  part of 's = inv(k) (m + xr) mod q' */
-    if ((kinv = BN_mod_inverse(NULL, k, dsa->q, ctx)) == NULL)
+    if ((kinv = dsa_inverse_mod(k, dsa->q, ctx)) == NULL)
```
Comment (mspncp) [inline crypto/dsa/dsa_ossl.c]: Just a curious question from a consistency addict: why the reordering from `mod_inverse` to `inverse_mod`?

File: crypto/dsa/dsa_ossl.c [hunk 2]
```diff
@@ -393,3 +394,30 @@ static int dsa_finish(DSA *dsa)
...
 }
+
+/*
+ * Compute the inverse of k modulo q.
+ * Since q is prime, Fermat's Little Theorem applies, which reduces this to
+ * mod-exp operation.  Both the exponent and modulus are public information
+ * so a constant time mod-exp shouldn't be required.  A newly allocated BIGNUM
+ * is returned which the caller must free.
+ */
+static BIGNUM *dsa_inverse_mod(const BIGNUM *k, const BIGNUM *q, BN_CTX *ctx)
```
Comment (mspncp) [inline crypto/dsa/dsa_ossl.c]: Also, wouldn't a name like `dsa_mod_inverse_fermat` be more expressive than just `dsa_inverse_mod`?

**Comment by mspncp** [comment]
BTW: should this go into 1.1.1, too?

**Comment by mspncp** [comment]
Approved for 1.1.1 and master.

File: crypto/dsa/dsa_ossl.c [hunk 3]
```diff
@@ -393,3 +394,30 @@ static int dsa_finish(DSA *dsa)
     BN_MONT_CTX_free(dsa->method_mont_p);
     return 1;
 }
+
+/*
+ * Compute the inverse of k modulo q.
+ * Since q is prime, Fermat's Little Theorem applies, which reduces this to
+ * mod-exp operation.  Both the exponent and modulus are public information
+ * so a constant time mod-exp shouldn't be required.  A newly allocated BIGNUM
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply crypto/dsa/dsa_ossl.c]: Oh, "m is even and the flag is set", I missed the "and" part.

File: crypto/dsa/dsa_ossl.c (hunk 2 above)
Comment (paulidale) [reply crypto/dsa/dsa_ossl.c]: Yes, I'll change this.

File: crypto/dsa/dsa_ossl.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidben) [reply crypto/dsa/dsa_ossl.c]: `BN_mod_exp_mont_consttime` (fixed-window, Montgomery reduction) is for secret exponents. The exponent is public. It's the base (the value being inverted) that is secret. There, `BN_mod_exp_mont` (variable-window, Montgomery reduction) is the correct exponentiation algorithm.

Remember, "constant-time" is not a meaningful statement on its own. It's a question of which values are and aren't secret.

[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/dsa/dsa_ossl.c]: It will end up in `BN_mod_exp_mont` as written but I'll make it explicit.

**Comment by mattcaswell** [comment]
Shouldn't this also go to 1.1.0 and 1.0.2?

**Comment by paulidale** [comment]
> Shouldn't this also go to 1.1.0 and 1.0.2?

Yes.

**Comment by paulidale** [comment]
Merged to master, 1.1.1 and 1.1.0.
1.0.2 in separate PR.

**Comment by rvagg** [comment]
Is this distinct from CVE-2018-0734? And does this one warrant a CVE?

**Comment by paulidale** [comment]
This one is different.  We didn't rate it as warranting a CVE.

**Pull request closed without merging**